File: UniLayoutIOS/UniLayout/Classes/containers/UniLinearContainer.swift
//
//  UniLinearContainer.swift
//  UniLayout Pod
//
//  Library container: a vertically or horizontally aligned view container
//  Stacks views below or to the right of each other
//

import UIKit

/// Specifies the layout direction of the subviews of the linear container
public enum UniLinearContainerOrientation: String {
    
    case vertical = "vertical"
    case horizontal = "horizontal"
    
}

/// A layout container view used to align subviews horizontally or vertically
open class UniLinearContainer: UIView, UniLayoutView, UniLayoutPaddedView {

    // ---
    // MARK: Layout integration
    // ---
    
    public var layoutProperties = UniLayoutProperties()
    public var padding = UIEdgeInsets.zero
    
    public var visibility: UniVisibility {
        set {
            isHidden = newValue != .visible
            layoutProperties.hiddenTakesSpace = newValue == .invisible
        }
        get {
            if isHidden {
                return layoutProperties.hiddenTakesSpace ? .invisible : .hidden
            }
            return .visible
        }
    }

    
    // ---
    // MARK: Members
    // ---
    
    public var orientation = UniLinearContainerOrientation.vertical

    
    // ---
    // MARK: Custom layout
    // ---

    @discardableResult internal func performVerticalLayout(sizeSpec: CGSize, widthSpec: UniMeasureSpec, heightSpec: UniMeasureSpec, adjustFrames: Bool) -> CGSize {
        // Determine available size without padding
        var paddedSize = CGSize(width: max(0, sizeSpec.width - padding.left - padding.right), height: max(0, sizeSpec.height - padding.top - padding.bottom))
        var measuredSize = CGSize(width: padding.left, height: padding.top)
        if widthSpec == .unspecified {
            paddedSize.width = 0xFFFFFF
        }
        if heightSpec == .unspecified {
            paddedSize.height = 0xFFFFFF
        }

        // Measure the views without any weight
        var subviewSizes: [CGSize] = []
        var totalWeight: CGFloat = 0
        var remainingHeight = paddedSize.height
        var totalMinHeightForWeight: CGFloat = 0
        for view in subviews {
            // Skip hidden views if they are not part of the layout
            if view.isHidden && !((view as? UniLayoutView)?.layoutProperties.hiddenTakesSpace ?? false) {
                subviewSizes.append(CGSize.zero)
                continue
            }

            // Skip views with weight, they will go in the second phase
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                if viewLayoutProperties.weight > 0 {
                    totalWeight += viewLayoutProperties.weight
                    remainingHeight -= viewLayoutProperties.margin.top + viewLayoutProperties.margin.bottom + viewLayoutProperties.minHeight
                    totalMinHeightForWeight += viewLayoutProperties.minHeight
                    subviewSizes.append(CGSize.zero)
                    continue
                }
            }
            
            // Perform measure and update remaining height
            var limitWidth = paddedSize.width
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                limitWidth -= viewLayoutProperties.margin.left + viewLayoutProperties.margin.right
                remainingHeight -= viewLayoutProperties.margin.top + viewLayoutProperties.margin.bottom
            }
            let result = UniLayout.measure(view: view, sizeSpec: CGSize(width: limitWidth, height: remainingHeight), parentWidthSpec: widthSpec, parentHeightSpec: heightSpec, forceViewWidthSpec: .unspecified, forceViewHeightSpec: .unspecified)
            remainingHeight = max(0, remainingHeight - result.height)
            subviewSizes.append(result)
        }
        
        // Measure the remaining views with weight
        remainingHeight += totalMinHeightForWeight
        for i in 0..<min(subviews.count, subviewSizes.count) {
            // Skip hidden views if they are not part of the layout
            let view = subviews[i]
            if view.isHidden && !((view as? UniLayoutView)?.layoutProperties.hiddenTakesSpace ?? false) {
                continue
            }
            
            // Continue with views with weight
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                if viewLayoutProperties.weight > 0 {
                    let forceViewHeightSpec: UniMeasureSpec = heightSpec == .exactSize ? .exactSize : .unspecified
                    let result = UniLayout.measure(view: view, sizeSpec: CGSize(width: paddedSize.width - viewLayoutProperties.margin.left - viewLayoutProperties.margin.right, height: remainingHeight * viewLayoutProperties.weight / totalWeight), parentWidthSpec: widthSpec, parentHeightSpec: heightSpec, forceViewWidthSpec: .unspecified, forceViewHeightSpec: forceViewHeightSpec)
                    remainingHeight = max(0, remainingHeight - result.height)
                    totalWeight -= viewLayoutProperties.weight
                    subviewSizes[i] = result
                }
            }
        }
        
        // Start doing layout
        var y = padding.top
        for i in 0..<min(subviews.count, subviewSizes.count) {
            // Skip hidden views if they are not part of the layout
            let view = subviews[i]
            if view.isHidden && !((view as? UniLayoutView)?.layoutProperties.hiddenTakesSpace ?? false) {
                continue
            }
            
            // Continue with the others
            let size = subviewSizes[i]
            var x = padding.left
            var nextY = y
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                x += viewLayoutProperties.margin.left
                y += viewLayoutProperties.margin.top
                if adjustFrames {
                    x += (paddedSize.width - viewLayoutProperties.margin.left - viewLayoutProperties.margin.right - size.width) * viewLayoutProperties.horizontalGravity
                }
                measuredSize.width = max(measuredSize.width, x + size.width + viewLayoutProperties.margin.right)
                nextY = y + size.height + viewLayoutProperties.margin.bottom
            } else {
                measuredSize.width = max(measuredSize.width, x + size.width)
                nextY = y + size.height
            }
            if adjustFrames {
                UniLayout.setFrame(view: view, frame: CGRect(x: x, y: y, width: size.width, height: size.height))
            }
            measuredSize.height = max(measuredSize.height, nextY)
            y = nextY
        }

        // Adjust final measure with padding and limitations
        measuredSize.width += padding.right
        measuredSize.height += padding.bottom
        if widthSpec == .exactSize {
            measuredSize.width = sizeSpec.width
        } else if widthSpec == .limitSize {
            measuredSize.width = min(measuredSize.width, sizeSpec.width)
        }
        if heightSpec == .exactSize {
            measuredSize.height = sizeSpec.height
        } else if heightSpec == .limitSize {
            measuredSize.height = min(measuredSize.height, sizeSpec.height)
        }
        return measuredSize
    }

    @discardableResult internal func performHorizontalLayout(sizeSpec: CGSize, widthSpec: UniMeasureSpec, heightSpec: UniMeasureSpec, adjustFrames: Bool) -> CGSize {
        // Determine available size without padding
        var paddedSize = CGSize(width: max(0, sizeSpec.width - padding.left - padding.right), height: max(0, sizeSpec.height - padding.top - padding.bottom))
        var measuredSize = CGSize(width: padding.left, height: padding.top)
        if widthSpec == .unspecified {
            paddedSize.width = 0xFFFFFF
        }
        if heightSpec == .unspecified {
            paddedSize.height = 0xFFFFFF
        }
        
        // Measure the views without any weight
        var subviewSizes: [CGSize] = []
        var totalWeight: CGFloat = 0
        var remainingWidth = paddedSize.width
        var totalMinWidthForWeight: CGFloat = 0
        for view in subviews {
            // Skip hidden views if they are not part of the layout
            if view.isHidden && !((view as? UniLayoutView)?.layoutProperties.hiddenTakesSpace ?? false) {
                subviewSizes.append(CGSize.zero)
                continue
            }
            
            // Skip views with weight, they will go in the second phase
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                if viewLayoutProperties.weight > 0 {
                    totalWeight += viewLayoutProperties.weight
                    remainingWidth -= viewLayoutProperties.margin.left + viewLayoutProperties.margin.right + viewLayoutProperties.minWidth
                    totalMinWidthForWeight += viewLayoutProperties.minWidth
                    subviewSizes.append(CGSize.zero)
                    continue
                }
            }
            
            // Perform measure and update remaining width
            var limitHeight = paddedSize.height
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                remainingWidth -= viewLayoutProperties.margin.left + viewLayoutProperties.margin.right
                limitHeight -= viewLayoutProperties.margin.top + viewLayoutProperties.margin.bottom
            }
            let result = UniLayout.measure(view: view, sizeSpec: CGSize(width: remainingWidth, height: limitHeight), parentWidthSpec: widthSpec, parentHeightSpec: heightSpec, forceViewWidthSpec: .unspecified, forceViewHeightSpec: .unspecified)
            remainingWidth = max(0, remainingWidth - result.width)
            subviewSizes.append(result)
        }
        
        // Measure the remaining views with weight
        remainingWidth += totalMinWidthForWeight
        for i in 0..<min(subviews.count, subviewSizes.count) {
            // Skip hidden views if they are not part of the layout
            let view = subviews[i]
            if view.isHidden && !((view as? UniLayoutView)?.layoutProperties.hiddenTakesSpace ?? false) {
                continue
            }
            
            // Continue with views with weight
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                if viewLayoutProperties.weight > 0 {
                    let forceViewWidthSpec: UniMeasureSpec = widthSpec == .exactSize ? .exactSize : .unspecified
                    let result = UniLayout.measure(view: view, sizeSpec: CGSize(width: remainingWidth * viewLayoutProperties.weight / totalWeight, height: paddedSize.height - viewLayoutProperties.margin.top - viewLayoutProperties.margin.bottom), parentWidthSpec: widthSpec, parentHeightSpec: heightSpec, forceViewWidthSpec: forceViewWidthSpec, forceViewHeightSpec: .unspecified)
                    remainingWidth = max(0, remainingWidth - result.width)
                    totalWeight -= viewLayoutProperties.weight
                    subviewSizes[i] = result
                }
            }
        }
        
        // Start doing layout
        var x = padding.left
        for i in 0..<min(subviews.count, subviewSizes.count) {
            // Skip hidden views if they are not part of the layout
            let view = subviews[i]
            if view.isHidden && !((view as? UniLayoutView)?.layoutProperties.hiddenTakesSpace ?? false) {
                continue
            }
            
            // Continue with the others
            let size = subviewSizes[i]
            var y = padding.top
            var nextX = x
            if let viewLayoutProperties = (view as? UniLayoutView)?.layoutProperties {
                x += viewLayoutProperties.margin.left
                y += viewLayoutProperties.margin.top
                if adjustFrames {
                    y += (paddedSize.height - viewLayoutProperties.margin.top - viewLayoutProperties.margin.bottom - size.height) * viewLayoutProperties.verticalGravity
                }
                measuredSize.height = max(measuredSize.height, y + size.height + viewLayoutProperties.margin.bottom)
                nextX = x + size.width + viewLayoutProperties.margin.right
            } else {
                measuredSize.height = max(measuredSize.height, y + size.height)
                nextX = x + size.width
            }
            if adjustFrames {
                UniLayout.setFrame(view: view, frame: CGRect(x: x, y: y, width: size.width, height: size.height))
            }
            measuredSize.width = max(measuredSize.width, nextX)
            x = nextX
        }
        
        // Adjust final measure with padding and limitations
        measuredSize.width += padding.right
        measuredSize.height += padding.bottom
        if widthSpec == .exactSize {
            measuredSize.width = sizeSpec.width
        } else if widthSpec == .limitSize {
            measuredSize.width = min(measuredSize.width, sizeSpec.width)
        }
        if heightSpec == .exactSize {
            measuredSize.height = sizeSpec.height
        } else if heightSpec == .limitSize {
            measuredSize.height = min(measuredSize.height, sizeSpec.height)
        }
        return measuredSize
    }

    open func measuredSize(sizeSpec: CGSize, widthSpec: UniMeasureSpec, heightSpec: UniMeasureSpec) -> CGSize {
        if orientation == .vertical {
            return performVerticalLayout(sizeSpec: sizeSpec, widthSpec: widthSpec, heightSpec: heightSpec, adjustFrames: false)
        }
        return performHorizontalLayout(sizeSpec: sizeSpec, widthSpec: widthSpec, heightSpec: heightSpec, adjustFrames: false)
    }
    
    open override func layoutSubviews() {
        if orientation == .vertical {
            performVerticalLayout(sizeSpec: bounds.size, widthSpec: .exactSize, heightSpec: .exactSize, adjustFrames: true)
        } else {
            performHorizontalLayout(sizeSpec: bounds.size, widthSpec: .exactSize, heightSpec: .exactSize, adjustFrames: true)
        }
    }
    
    open override func systemLayoutSizeFitting(_ targetSize: CGSize, withHorizontalFittingPriority horizontalFittingPriority: UILayoutPriority, verticalFittingPriority: UILayoutPriority) -> CGSize {
        return measuredSize(sizeSpec: targetSize, widthSpec: horizontalFittingPriority == UILayoutPriorityRequired ? UniMeasureSpec.limitSize : UniMeasureSpec.unspecified, heightSpec: verticalFittingPriority == UILayoutPriorityRequired ? UniMeasureSpec.limitSize : UniMeasureSpec.unspecified)
    }
 
    
    // ---
    // MARK: Improve layout needed behavior
    // ---

    open override func willRemoveSubview(_ subview: UIView) {
        UniLayout.setNeedsLayout(view: self)
    }
    
    open override func didAddSubview(_ subview: UIView) {
        UniLayout.setNeedsLayout(view: self)
    }

}

File: UniLayoutAndroid/UniLayoutLib/src/main/java/com/crescentflare/unilayout/containers/UniLinearContainer.java
package com.crescentflare.unilayout.containers;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Point;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.crescentflare.unilayout.R;
import com.crescentflare.unilayout.helpers.UniLayoutParams;
import com.crescentflare.unilayout.views.UniView;

import java.util.ArrayList;
import java.util.List;

/**
 * UniLayout container: a vertically or horizontally aligned view container
 * Stacks views below or to the right of each other
 */
public class UniLinearContainer extends ViewGroup
{
    // ---
    // Constants
    // ---

    public static final int VERTICAL = LinearLayout.VERTICAL;
    public static final int HORIZONTAL = LinearLayout.HORIZONTAL;


    // ---
    // Members
    // ---

    int orientation = VERTICAL;


    // ---
    // Initialization
    // ---

    public UniLinearContainer(Context context)
    {
        this(context, (AttributeSet)null);
    }

    public UniLinearContainer(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        init(attrs);
    }

    public UniLinearContainer(Context context, AttributeSet attrs, int defStyleAttr)
    {
        this(context, attrs);
    }

    public UniLinearContainer(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes)
    {
        this(context, attrs);
    }

    private void init(AttributeSet attrs)
    {
        if (attrs != null)
        {
            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.UniLinearContainer);
            orientation = a.getInt(R.styleable.UniLinearContainer_uni_orientation, VERTICAL);
            a.recycle();
        }
    }

    @Override
    protected LayoutParams generateDefaultLayoutParams()
    {
        return new UniLayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
    }

    @Override
    public LayoutParams generateLayoutParams(AttributeSet attrs)
    {
        return new UniLayoutParams(getContext(), attrs);
    }


    // ---
    // Adjust properties
    // ---

    public int getOrientation()
    {
        return orientation;
    }

    public void setOrientation(int orientation)
    {
        this.orientation = orientation;
    }


    // ---
    // Custom layout
    // ---

    private Point measuredSize = new Point();

    private void performVerticalLayout(Point measuredSize, int widthSize, int heightSize, int widthSpec, int heightSpec, boolean adjustLayout)
    {
        // Determine available size without padding
        int paddedWidthSize = widthSize - getPaddingLeft() - getPaddingRight();
        int paddedHeightSize = heightSize - getPaddingTop() - getPaddingBottom();
        measuredSize.x = getPaddingLeft();
        measuredSize.y = getPaddingTop();
        if (widthSpec == MeasureSpec.UNSPECIFIED)
        {
            paddedWidthSize = 0xFFFFFF;
        }
        if (heightSpec == MeasureSpec.UNSPECIFIED)
        {
            paddedHeightSize = 0xFFFFFF;
        }

        // Measure the views without any weight
        float totalWeight = 0;
        int remainingHeight = paddedHeightSize;
        float totalMinHeightForWeight = 0;
        int childCount = getChildCount();
        for (int i = 0; i < childCount; i++)
        {
            // Skip hidden views if they are not part of the layout
            View view = getChildAt(i);
            if (view.getVisibility() == GONE)
            {
                continue;
            }

            // Skip views with weight, they will go in the second phase
            ViewGroup.LayoutParams viewLayoutParams = view.getLayoutParams();
            if (viewLayoutParams instanceof UniLayoutParams)
            {
                UniLayoutParams uniLayoutParams = (UniLayoutParams)viewLayoutParams;
                if (uniLayoutParams.weight > 0)
                {
                    totalWeight += uniLayoutParams.weight;
                    remainingHeight -= uniLayoutParams.topMargin + uniLayoutParams.bottomMargin + uniLayoutParams.minHeight;
                    totalMinHeightForWeight += uniLayoutParams.minHeight;
                    continue;
                }
            }

            // Perform measure and update remaining height
            int limitWidth = paddedWidthSize;
            if (viewLayoutParams instanceof MarginLayoutParams)
            {
                limitWidth -= ((MarginLayoutParams)viewLayoutParams).leftMargin + ((MarginLayoutParams)viewLayoutParams).rightMargin;
                remainingHeight -= ((MarginLayoutParams)viewLayoutParams).topMargin + ((MarginLayoutParams)viewLayoutParams).bottomMargin;
            }
            UniView.uniMeasure(view, limitWidth, remainingHeight, widthSpec, heightSpec, MeasureSpec.UNSPECIFIED, MeasureSpec.UNSPECIFIED);
            remainingHeight = Math.max(0, remainingHeight - view.getMeasuredHeight());
        }

        // Measure the remaining views with weight
        remainingHeight += totalMinHeightForWeight;
        for (int i = 0; i < childCount; i++)
        {
            // Skip hidden views if they are not part of the layout
            View view = getChildAt(i);
            if (view.getVisibility() == GONE)
            {
                continue;
            }

            // Continue measuring views with weight
            if (view.getLayoutParams() instanceof UniLayoutParams)
            {
                UniLayoutParams uniLayoutParams = (UniLayoutParams)view.getLayoutParams();
                if (uniLayoutParams.weight > 0)
                {
                    int forceViewHeightSpec = heightSpec == MeasureSpec.EXACTLY ? MeasureSpec.EXACTLY : MeasureSpec.UNSPECIFIED;
                    UniView.uniMeasure(view, paddedWidthSize - uniLayoutParams.leftMargin - uniLayoutParams.rightMargin, (int)(remainingHeight * uniLayoutParams.weight / totalWeight), widthSpec, heightSpec, MeasureSpec.UNSPECIFIED, forceViewHeightSpec);
                    remainingHeight = Math.max(0, remainingHeight - view.getMeasuredHeight());
                    totalWeight -= uniLayoutParams.weight;
                }
            }
        }

        // Start doing layout
        int y = getPaddingTop();
        for (int i = 0; i < childCount; i++)
        {
            // Skip hidden views if they are not part of the layout
            View view = getChildAt(i);
            if (view.getVisibility() == GONE)
            {
                continue;
            }

            // Continue with the others
            int width = view.getMeasuredWidth();
            int height = view.getMeasuredHeight();
            int x = getPaddingLeft();
            int nextY = y;
            if (view.getLayoutParams() instanceof UniLayoutParams)
            {
                UniLayoutParams uniLayoutParams = (UniLayoutParams)view.getLayoutParams();
                x += uniLayoutParams.leftMargin;
                y += uniLayoutParams.topMargin;
                if (adjustLayout)
                {
                    x += (paddedWidthSize - uniLayoutParams.leftMargin - uniLayoutParams.rightMargin - width) * uniLayoutParams.horizontalGravity;
                }
                measuredSize.x = Math.max(measuredSize.x, x + width + uniLayoutParams.rightMargin);
                nextY = y + height + uniLayoutParams.bottomMargin;
            }
            else
            {
                measuredSize.x = Math.max(measuredSize.x, x + width);
                nextY = y + height;
            }
            if (adjustLayout)
            {
                view.layout(x, y, x + width, y + height);
            }
            measuredSize.y = Math.max(measuredSize.y, nextY);
            y = nextY;
        }

        // Adjust final measure with padding and limitations
        measuredSize.x += getPaddingRight();
        measuredSize.y += getPaddingBottom();
        if (widthSpec == MeasureSpec.EXACTLY)
        {
            measuredSize.x = widthSize;
        }
        else if (widthSpec == MeasureSpec.AT_MOST)
        {
            measuredSize.x = Math.min(measuredSize.x, widthSize);
        }
        if (heightSpec == MeasureSpec.EXACTLY)
        {
            measuredSize.y = heightSize;
        }
        else if (heightSpec == MeasureSpec.AT_MOST)
        {
            measuredSize.y = Math.min(measuredSize.y, heightSize);
        }
    }

    private void performHorizontalLayout(Point measuredSize, int widthSize, int heightSize, int widthSpec, int heightSpec, boolean adjustLayout)
    {
        // Determine available size without padding
        int paddedWidthSize = widthSize - getPaddingLeft() - getPaddingRight();
        int paddedHeightSize = heightSize - getPaddingTop() - getPaddingBottom();
        measuredSize.x = getPaddingLeft();
        measuredSize.y = getPaddingTop();
        if (widthSpec == MeasureSpec.UNSPECIFIED)
        {
            paddedWidthSize = 0xFFFFFF;
        }
        if (heightSpec == MeasureSpec.UNSPECIFIED)
        {
            paddedHeightSize = 0xFFFFFF;
        }

        // Measure the views without any weight
        float totalWeight = 0;
        int remainingWidth = paddedWidthSize;
        float totalMinWidthForWeight = 0;
        int childCount = getChildCount();
        for (int i = 0; i < childCount; i++)
        {
            // Skip hidden views if they are not part of the layout
            View view = getChildAt(i);
            if (view.getVisibility() == GONE)
            {
                continue;
            }

            // Skip views with weight, they will go in the second phase
            ViewGroup.LayoutParams viewLayoutParams = view.getLayoutParams();
            if (viewLayoutParams instanceof UniLayoutParams)
            {
                UniLayoutParams uniLayoutParams = (UniLayoutParams)viewLayoutParams;
                if (uniLayoutParams.weight > 0)
                {
                    totalWeight += uniLayoutParams.weight;
                    remainingWidth -= uniLayoutParams.leftMargin + uniLayoutParams.rightMargin + uniLayoutParams.minWidth;
                    totalMinWidthForWeight += uniLayoutParams.minWidth;
                    continue;
                }
            }

            // Perform measure and update remaining height
            int limitHeight = paddedHeightSize;
            if (viewLayoutParams instanceof MarginLayoutParams)
            {
                remainingWidth -= ((MarginLayoutParams)viewLayoutParams).leftMargin + ((MarginLayoutParams)viewLayoutParams).rightMargin;
                limitHeight -= ((MarginLayoutParams)viewLayoutParams).topMargin + ((MarginLayoutParams)viewLayoutParams).bottomMargin;
            }
            UniView.uniMeasure(view, remainingWidth, limitHeight, widthSpec, heightSpec, MeasureSpec.UNSPECIFIED, MeasureSpec.UNSPECIFIED);
            remainingWidth = Math.max(0, remainingWidth - view.getMeasuredWidth());
        }

        // Measure the remaining views with weight
        remainingWidth += totalMinWidthForWeight;
        for (int i = 0; i < childCount; i++)
        {
            // Skip hidden views if they are not part of the layout
            View view = getChildAt(i);
            if (view.getVisibility() == GONE)
            {
                continue;
            }

            // Continue with views with weight
            if (view.getLayoutParams() instanceof UniLayoutParams)
            {
                UniLayoutParams uniLayoutParams = (UniLayoutParams)view.getLayoutParams();
                if (uniLayoutParams.weight > 0)
                {
                    int forceViewWidthSpec = widthSpec == MeasureSpec.EXACTLY ? MeasureSpec.EXACTLY : MeasureSpec.UNSPECIFIED;
                    UniView.uniMeasure(view, (int)(remainingWidth * uniLayoutParams.weight / totalWeight), paddedHeightSize - uniLayoutParams.topMargin - uniLayoutParams.bottomMargin, widthSpec, heightSpec, forceViewWidthSpec, MeasureSpec.UNSPECIFIED);
                    remainingWidth = Math.max(0, remainingWidth - view.getMeasuredWidth());
                    totalWeight -= uniLayoutParams.weight;
                }
            }
        }

        // Start doing layout
        int x = getPaddingLeft();
        for (int i = 0; i < childCount; i++)
        {
            // Skip hidden views if they are not part of the layout
            View view = getChildAt(i);
            if (view.getVisibility() == GONE)
            {
                continue;
            }

            // Continue with the others
            int width = view.getMeasuredWidth();
            int height = view.getMeasuredHeight();
            int y = getPaddingTop();
            int nextX = x;
            if (view.getLayoutParams() instanceof UniLayoutParams)
            {
                UniLayoutParams uniLayoutParams = (UniLayoutParams)view.getLayoutParams();
                x += uniLayoutParams.leftMargin;
                y += uniLayoutParams.topMargin;
                if (adjustLayout)
                {
                    y += (paddedHeightSize - uniLayoutParams.topMargin - uniLayoutParams.bottomMargin - height) * uniLayoutParams.verticalGravity;
                }
                measuredSize.y = Math.max(measuredSize.y, y + height + uniLayoutParams.bottomMargin);
                nextX = x + width + uniLayoutParams.rightMargin;
            }
            else
            {
                measuredSize.y = Math.max(measuredSize.y, y + height);
                nextX = x + width;
            }
            if (adjustLayout)
            {
                view.layout(x, y, x + width, y + height);
            }
            measuredSize.x = Math.max(measuredSize.x, nextX);
            x = nextX;
        }

        // Adjust final measure with padding and limitations
        measuredSize.x += getPaddingRight();
        measuredSize.y += getPaddingBottom();
        if (widthSpec == MeasureSpec.EXACTLY)
        {
            measuredSize.x = widthSize;
        }
        else if (widthSpec == MeasureSpec.AT_MOST)
        {
            measuredSize.x = Math.min(measuredSize.x, widthSize);
        }
        if (heightSpec == MeasureSpec.EXACTLY)
        {
            measuredSize.y = heightSize;
        }
        else if (heightSpec == MeasureSpec.AT_MOST)
        {
            measuredSize.y = Math.min(measuredSize.y, heightSize);
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
    {
        if (getOrientation() == VERTICAL)
        {
            performVerticalLayout(measuredSize, MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.getSize(heightMeasureSpec), MeasureSpec.getMode(widthMeasureSpec), MeasureSpec.getMode(heightMeasureSpec), false);
            setMeasuredDimension(measuredSize.x, measuredSize.y);
        }
        else
        {
            performHorizontalLayout(measuredSize, MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.getSize(heightMeasureSpec), MeasureSpec.getMode(widthMeasureSpec), MeasureSpec.getMode(heightMeasureSpec), false);
            setMeasuredDimension(measuredSize.x, measuredSize.y);
        }
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom)
    {
        if (getOrientation() == VERTICAL)
        {
            performVerticalLayout(measuredSize, right - left, bottom - top, MeasureSpec.EXACTLY, MeasureSpec.EXACTLY, true);
        }
        else
        {
            performHorizontalLayout(measuredSize, right - left, bottom - top, MeasureSpec.EXACTLY, MeasureSpec.EXACTLY, true);
        }
    }
}
